shared-bindings/alarm/pin/PinAlarm.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Dan Halbert for Adafruit Industries
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/board/__init__.h"
+#include "shared-bindings/microcontroller/__init__.h"
+#include "shared-bindings/microcontroller/Pin.h"
+#include "shared-bindings/alarm/pin/PinAlarm.h"
+
+#include "py/nlr.h"
+#include "py/obj.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "supervisor/shared/translate.h"
+
+//| class PinAlarm:
+//|     """Trigger an alarm when a pin changes state."""
+//|
+//|     def __init__(self, *pins: microcontroller.Pin, value: bool, all_same_value: bool = False, edge: bool = False, pull: bool = False) -> None:
+//|         """Create an alarm triggered by a `microcontroller.Pin` level. The alarm is not active
+//|         until it is passed to an `alarm`-enabling function, such as `alarm.sleep_until_alarms()` or
+//|         `alarm.exit_and_deep_sleep_until_alarms()`.
+//|
+//|         :param microcontroller.Pin \*pins: The pins to monitor. On some ports, the choice of pins
+//|           may be limited due to hardware restrictions, particularly for deep-sleep alarms.
```

Comment:
I don't think you should have multiple pin support to start. It's too ESP specific and not likely needed.

(Remove `all_same_value` too.)

Reply:
OK. Other chips have various idiosyncratic wakeup mechanisms.

nRF allows an OR'ing of pin DETECT signals (detect low or high). It also has a latch that detects whether the enabled pins were detected, though not necessarily simultaneously. So, e.g., it could wake up when you press buttons 1,2,3, but they don't have to be simultaneous. 

SAMD51 has tamper-detect pins with debouncing.

STM is more confusing.